server/src/main/java/org/opensearch/action/search/SearchRequestStats.java

```diff
@@ -0,0 +1,137 @@
+ */
+
+package org.opensearch.action.search;
+
+import org.opensearch.common.inject.Inject;
+import org.opensearch.common.metrics.CounterMetric;
+import org.opensearch.common.metrics.MeanMetric;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+/**
+ * Request level search stats to track coordinator level node search latencies
+ *
+ * @opensearch.internal
+ */
+public final class SearchRequestStats implements SearchRequestOperationsListener {
+    public StatsHolder totalStats = new StatsHolder();
+
+    ConcurrentHashMap<SearchPhaseName, Consumer<SearchPhaseContext>> phaseEndTracker = new ConcurrentHashMap<>();
+    ConcurrentHashMap<SearchPhaseName, Consumer<SearchPhaseContext>> phaseStartTracker = new ConcurrentHashMap<>();
+    ConcurrentHashMap<SearchPhaseName, Consumer<SearchPhaseContext>> phaseFailureTracker = new ConcurrentHashMap<>();
+
+    @Inject
+    public SearchRequestStats() {
+        for (SearchPhaseName searchPhase : SearchPhaseName.values()) {
+            phaseEndTracker.put(searchPhase, searchPhaseContext -> {
+                if (SearchPhaseName.DFS_QUERY.equals(SearchPhaseName.getSearchPhaseName(searchPhaseContext.getCurrentPhase().getName()))) {
+                    totalStats.queryCurrentMap.get(SearchPhaseName.QUERY).dec();
+                    totalStats.queryTotalMap.get(SearchPhaseName.QUERY).inc();
+                    totalStats.queryMetricMap.get(SearchPhaseName.QUERY)
+                        .inc(TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - searchPhaseContext.getCurrentPhase().getStartTimeInNanos()));
+                } else {
+                    totalStats.queryCurrentMap.get(searchPhase).dec();
+                    totalStats.queryTotalMap.get(searchPhase).inc();
+                    totalStats.queryMetricMap.get(searchPhase)
+                        .inc(TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - searchPhaseContext.getCurrentPhase().getStartTimeInNanos()));
+                }
+            });
+            phaseStartTracker.put(searchPhase, searchPhaseContext -> {
+                if (SearchPhaseName.DFS_QUERY.equals(SearchPhaseName.getSearchPhaseName(searchPhaseContext.getCurrentPhase().getName()))) {
+                    totalStats.queryCurrentMap.get(SearchPhaseName.QUERY).inc();
+                } else {
+                    totalStats.queryCurrentMap.get(searchPhase).inc();
+                }
+            });
+            phaseFailureTracker.put(searchPhase, searchPhaseContext -> {
+                if (SearchPhaseName.DFS_QUERY.equals(SearchPhaseName.getSearchPhaseName(searchPhaseContext.getCurrentPhase().getName()))) {
+                    totalStats.queryCurrentMap.get(SearchPhaseName.QUERY).dec();
+                } else {
+                    totalStats.queryCurrentMap.get(searchPhase).dec();
+                }
+            });
+        }
```

Comment:
This is unnecessarily complicated.

Just define a class to hold the stats for a single phase:
```
class StatsHolder {
  CounterMetric current;
  CounterMetric total;
  MeanMetric timing;
}
```

You need a single `Map<SearchPhaseName, StatsHolder> phaseStatsMap = new EnumMap(SearchPhaseName.class)`, as the sole member of `SearchRequestStats`.

In the constructor, initialize it for all the phases:

```
for (SearchPhaseName name: SearchPhaseName.values()) {
  phaseStatsMap.put(name, new StatsHolder());
}
```

You can implement the listener methods like:

```
public void onPhaseStart(SearchPhaseContext context) {
  // The following is guaranteed to be non-null because every possible phase had a StatsHolder initialized in the constructor.
  StatsHolder phaseStats = phaseStatsMap.get(context.getCurrentPhase().getSearchPhaseName());
  phaseStats.current.inc();
}

public void onPhaseEnd(SearchPhaseContext context) {
  StatsHolder phaseStats = phaseStatsMap.get(context.getCurrentPhase().getSearchPhaseName());
  phaseStats.current.dec();
  phaseStats.total.inc();
  phaseStats.timing.inc(TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - searchPhaseContext.getCurrentPhase().getStartTimeInNanos()));
}

public void onPhaseFailure(SearchPhaseContext context) {
  StatsHolder phaseStats = phaseStatsMap.get(context.getCurrentPhase().getSearchPhaseName());
  phaseStats.current.dec();
}
```

The individual counters are threadsafe, since they're essentially `LongAdder`s, and the content of `phaseStatsMap` doesn't change (i.e. there is no `put` or `remove` after the constructor), so it doesn't need to be a `ConcurrentMap`.

Regarding combining the DFS_QUERY and QUERY stats, I would leave that until later, if they need to be combined at all.

Reply:
Thanks @msfroh for the detailed explanation. This make sense! My bad for making it complicated.